doc/developer/design/20230519_statement_logging.md

```diff
@@ -0,0 +1,165 @@
+  persist the text of every statement (in either Cockroach or S3),
+  which we consider unacceptable for `SELECT` queries.
+
+## Overview
+
+The coordinator maintains a buffer of pending events related to
+statement execution. Whenever a prepared statement is created, a
+corresponding event is added to the buffer. Whenever a statement is
+executed, a system var called `STATEMENT_LOG_SAMPLING_RATE` is
+queried, and a random number between 0 and 1 is generated. If the
+random number is greater than the sampling rate, a description of the
+statement invocation is written into the buffer.
+
+Every so often (controlled by `STATEMENT_LOG_FLUSH_FREQUENCY`), the
+buffer is converted to updates that are written out to
+`mz_statement_execution_history` and `mz_prepared_statement_history`,
+with a column in the former referencing the latter. These system
+tables are persistent, but entries older than `STATEMENT_LOG_MAX_AGE`
+are periodically deleted.
+
+## Detailed description
+
+The coordinator maintains a buffer of pending events, which can be of
+the following variants:
+
+``` rust
+enum StatementLogEvent {
+    StatementPrepared {
+        id: Uuid,
+        sql: String,
+        name: Option<String>,
+    },
+    StatementExecutionBegan {
+        id: Uuid,
+        prepared_statement_id: Uuid,        
+        params: Vec<Datum>,
+        when: DateTime,
+        // Additional metadata might eventually
+        // go here, e.g. `plan: Option<OptimizedMirRelationExpr>`
+    },
+    StatementExecutionFinished {
+        id: Uuid,
+        when: DateTime,
+        n_results: Option<usize>,
+    },
+}
+```
+
+The `StatementPrepared` and `StatementExecutionBegan` events are
+generated and stored in the buffer in `fn handle_execute`, after
+checking a random value against `STATEMENT_LOG_SAMPLING_RATE` . This
+requires keeping track of the prepared statement bound to each name,
+and whether it has been logged yet (as we don't want to log each
+statement several times). The `StatementExecutionFinished` events are
+generated in several different places, depending on the type of
+statement. For example, the event for `SELECT` queries is generated in
+response to the `ControllerResponse::PeekResponse` message.
+
+The coordinator also holds on to an interval object whose tick rate is
+controlled by `STATEMENT_LOG_FLUSH_FREQUENCY`. It listens for ticks
+on that interval in its main `serve` loop; when such a tick occurs, it
+emits data to the system tables, as follows:
+
+* Each `StatementPrepared` event generates an entry in
+  `mz_prepared_statement_history`.
+* Each `StatementExecutionBegan` event generates an entry in
+  `mz_statement_execution_history`, and records that entry for
+  retraction as part of a later update.
+* Each `StatementExecutionEnded` event generates an update to the
+  entry described above.
+
+These entries are consolidated and then emitted to the builtin
+tables, and also to the catalog, ensuring durability.
```

Comment:
Can you spell out the schema of `mz_prepared_statement_history` and `mz_statement_execution_history`? Ideally in the format that we can later just copy/paste into the `mz_internal` reference docs. I'm curious about two things in particular:

1. How do we distinguish between begin and end events in `mz_statement_execution_history`?
2. How are you going to represent the `params` field in SQL?

Reply:
+1 this would be helpful to see the full list of columns to be provided in those tables.

Surfaces would like to use the data for at least a few use cases:
1) Analyze average and p99 latency for different types of queries. In addition to the query duration, we'd like to be able to slice it by a number of dimensions.
a) success vs failure
b) number of records returned
c) query type (DDL, DML, select) - we can always do regex matching on the raw sql if needed
d) whether an index was used (stretch goal)
2) Which functions and sql commands customers are actually using. We can do regex matching on the raw sql for this one, or just some manual scanning.
3) Understanding where users are hitting sql errors, and how often. This one will be manual exploration filtering on only queries that failed.

It seems like this design will definitely cover 1a, 1b, and 3, and probably 1c and 2 via regex matching.